File: src/Particle.hs
module Particle (
    Particle(..),
    particleMass,
    particleCharge
) where

data Particle = Electron | Nucleus Double Double Double deriving Eq

particleMass :: Particle -> Double
particleMass Electron = 1
particleMass (Nucleus _ _ m) = m

particleCharge :: Particle -> Double
particleCharge Electron = -1
particleCharge (Nucleus n _ _) = n

instance Show Particle where
    show Electron = "e"
    show (Nucleus z a m) = "Z"++show z++"A"++show a++(if m<1/0 then "m"++show m else "")

File: Rust/src/particle.rs
use std::fmt::{self, Display};

#[derive(Clone, PartialEq)]
pub enum Particle {
  Electron,
  Nucleus {z:f64, a:f64, m:f64},
}

pub fn particle_mass(particle : &Particle) -> f64 {
  match particle {
    Particle::Electron => 1.0,
    Particle::Nucleus{m,..} => *m,
  }
}

pub fn particle_charge(particle : &Particle) -> f64 {
  match particle {
    Particle::Electron => -1.0,
    Particle::Nucleus{z,..} => *z,
  }
}

impl Display for Particle {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    match &self {
      Particle::Electron => write!(f, "e"),
      Particle::Nucleus{z,a,m} => if m.is_infinite() {
          write!(f, "Z{}A{}", z, a)
        } else {
          write!(f, "Z{}A{}m{}", z, a, m)
        },
    }
  }
}
